### PR TITLE
layers: Hack to not have piglet exit() destroy maps under us

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Really just an attempt to showcase "a fix" for https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8618

### CODE REVIEW

File: layers/core_checks/cc_device.cpp [hunk 1]
```diff
@@ -1023,3 +1023,36 @@ bool CoreChecks::PreCallValidateCreatePipelineBinariesKHR(VkDevice device, const
...
+        // will crash and will get a false report it "crashes only with VVL" (Currently this is only required in CoreChecks)
+        exit_called = true;
+        printf("--- ~~~StaticAllocatedObject() --\n");
+        fflush(stdout);
+    }
+    volatile bool exit_called = false;
+};
+
+const StaticAllocatedObject &GetStaticAllocatedObject() {
+    static const StaticAllocatedObject object;
```
Comment (artem-lunarg) [inline layers/core_checks/cc_device.cpp:1045]: One thing to check if this is important for issue's use case. Constructor for static object in function scope runs the first time the function is called. If the function is called the first time after exit() is called then `exit_called` will stay false.

Comment (artem-lunarg) [reply layers/core_checks/cc_device.cpp:1045]: I'm a bit a afraid of this kind of stuff, if @jeremyg-lunarg is available to check it would be nice.

**Comment by jeremyg-lunarg** [comment]
Could you also do this with an https://en.cppreference.com/w/c/program/atexit callback?

**Comment by spencer-lunarg** [comment]
> Could you also do this with an https://en.cppreference.com/w/c/program/atexit callback?

the issue is it replays it back in the order they are created, so piglet does it sometime after `vkCreateDevice()`

so it ends looking like

```c++
vvl::CreateDevice(); // VVL hooks into atexit()
PigletSetup(); // piglet hooks into atexit()

// ... app

exit()

PigletDestroy(); // runs their atexit() function
// <----- Crash here
vvl::OnExit(); // We never get to run ours
```

File: layers/core_checks/cc_device.cpp [hunk 2]
```diff
@@ -1023,3 +1023,36 @@ bool CoreChecks::PreCallValidateCreatePipelineBinariesKHR(VkDevice device, const
...
+
+class StaticAllocatedObject {
+  public:
+    StaticAllocatedObject() {
+        printf("--- StaticAllocatedObject() --\n");
+        fflush(stdout);
+    };
+    ~StaticAllocatedObject() {
+        // This is used to detect if the program has called exit() and we need to cleanup the layers. If we don't the app might have
+        // a fucntion in atexit() trying to still make Vulkan calls and if it hits our freed heap memory (such as our std::map) it
```
Comment (filnet) [inline layers/core_checks/cc_device.cpp:1035]:
```suggestion
        // a function in atexit() trying to still make Vulkan calls and if it hits our freed heap memory (such as our std::map) it
```

File: layers/core_checks/cc_device.cpp (hunk 1 above)
Comment (artem-lunarg) [reply layers/core_checks/cc_device.cpp:1045]: Dll boundaries might create additional difficulties that are not necessarily covered by the standard, just something to pay attention too and to check platform/compiler specific behavior.

**Pull request closed without merging**